spec.html

```diff
@@ -42415,104 +42525,218 @@ <h1>
-          1. Let _len_ be the number of elements in _S_.
+          1. Let _len_ be the number of elements in _WL_.[[Waiters]].
           1. Let _n_ be min(_c_, _len_).
-          1. Let _L_ be a List whose elements are the first _n_ elements of _S_.
-          1. Remove the first _n_ elements of _S_.
+          1. Let _L_ be a List whose elements are the first _n_ elements of _WL_.[[Waiters]].
+          1. Remove the first _n_ elements of _WL_.[[Waiters]].
           1. Return _L_.
         </emu-alg>
       </emu-clause>
 
-      <emu-clause id="sec-suspendagent" type="abstract operation" oldids="sec-suspend">
+      <emu-clause id="sec-suspendthisagent" type="abstract operation" oldids="sec-suspend,sec-suspendagent">
         <h1>
-          SuspendAgent (
-            _WL_: a WaiterList,
-            _W_: an agent signifier,
-            _minimumTimeout_: a non-negative extended mathematical value,
-          ): a Boolean
+          SuspendThisAgent (
+            _WL_: a WaiterList Record,
+            _waiterRecord_: a Waiter Record,
+          ): ~unused~
         </h1>
         <dl class="header">
         </dl>
         <emu-alg>
           1. Assert: The surrounding agent is in the critical section for _WL_.
-          1. Assert: _W_ is AgentSignifier().
-          1. Assert: _W_ is on the list of waiters in _WL_.
+          1. Assert: _WL_.[[Waiters]] contains _waiterRecord_.
+          1. Let _thisAgent_ be AgentSignifier().
+          1. Assert: _waiterRecord_.[[AgentSignifier]] is _thisAgent_ and _waiterRecord_.[[PromiseCapability]] is ~blocking~.
           1. Assert: AgentCanSuspend() is *true*.
-          1. Let _additionalTimeout_ be an implementation-defined non-negative mathematical value.
-          1. Let _timeout_ be _minimumTimeout_ + _additionalTimeout_.
-          1. NOTE: When _minimumTimeout_ is +∞, _timeout_ is also +∞ and the following step can terminate only by another agent calling NotifyWaiter.
-          1. Perform LeaveCriticalSection(_WL_) and suspend _W_ for up to _timeout_ milliseconds, performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. _W_ can wake from suspension either because the timeout expired or because it was notified explicitly by another agent calling NotifyWaiter with arguments _WL_ and _W_, and not for any other reasons at all.
+          1. Perform LeaveCriticalSection(_WL_) and suspend the surrounding agent until the time is _waiterRecord_.[[TimeoutTime]], performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. The surrounding agent can wake from suspension either because the timeout expired or because it was notified explicitly by another agent calling NotifyWaiter with arguments _WL_ and _thisAgent_, and not for any other reasons at all.
```

Comment:
The case of "another agent calling NotifyWaiter" now also includes the case where the timeout expired, which maybe calls for some rewording, e.g.:

`The only way for the surrounding agent to wake from suspension is by another agent calling NotifyWaiter with arguments _WL_ and _thisAgent_. This can happen if the timeout expires, or via a call to Atomics.notify.`

Reply:
Good suggestion.